pkg/controller/awsfederatedaccountaccess/awsfederatedaccountaccess_controller_test.go

```diff
@@ -0,0 +1,351 @@
+					awsOutputPolicy,
+					nil,
+				)
+			}
+
+			r := ReconcileAWSFederatedAccountAccess{}
+
+			afr := awsv1alpha1.AWSFederatedRole{
+				Spec: awsv1alpha1.AWSFederatedRoleSpec{
+					AWSCustomPolicy: newTestAwsCustomPolicyBuilder().awsCustomPol},
+			}
+
+			afaa := awsv1alpha1.AWSFederatedAccountAccess{
+				ObjectMeta: v1.ObjectMeta{
+					Labels: test.uidLabel,
+				}}
+
+			createPolicyOutput, err := r.createIAMPolicy(mockAWSClient, afr, afaa)
+			if test.expectedErr != nil {
+				assert.Equal(t, err, test.expectedErr)
+			}
+			assert.Equal(t, createPolicyOutput, test.createIAMPolicyOutput)
+		})
+	}
+
+}
+
+func TestCreateIAMRole(t *testing.T) {
+
+	awsOutputRole := &iam.CreateRoleOutput{
+		Role: &iam.Role{},
+	}
+
+	tests := []struct {
+		name                string
+		uidLabel            map[string]string
+		createIAMRoleOutput *iam.Role
+		expectedErr         error
+	}{
+		{
+			name:                "Test for UID label present",
+			uidLabel:            map[string]string{"uid": "abcd"},
+			createIAMRoleOutput: awsOutputRole.Role,
+			expectedErr:         nil,
+		},
+		{
+			name:                "Test for UID label missing",
+			uidLabel:            map[string]string{},
+			createIAMRoleOutput: nil,
+			expectedErr:         errors.New("Failed to get UID label"),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+
+			mocks := setupDefaultMocks(t, []runtime.Object{})
+
+			mockAWSClient := mock.NewMockClient(mocks.mockCtrl)
+
+			defer mocks.mockCtrl.Finish()
+
+			if test.createIAMRoleOutput != nil {
+				mockAWSClient.EXPECT().CreateRole(gomock.Any()).Return(
+					awsOutputRole,
+					nil,
+				)
+			}
+
+			r := ReconcileAWSFederatedAccountAccess{}
+
+			afr := awsv1alpha1.AWSFederatedRole{
+				ObjectMeta: v1.ObjectMeta{
+					Name: "",
+				},
+				Spec: awsv1alpha1.AWSFederatedRoleSpec{
+					AWSCustomPolicy: newTestAwsCustomPolicyBuilder().awsCustomPol},
+			}
+
+			afaa := awsv1alpha1.AWSFederatedAccountAccess{
+				ObjectMeta: v1.ObjectMeta{
+					Labels: test.uidLabel},
+				Spec: awsv1alpha1.AWSFederatedAccountAccessSpec{
+					ExternalCustomerAWSIAMARN: "",
+				},
+			}
+
+			createRoleOutput, err := r.createIAMRole(mockAWSClient, afr, afaa)
+			if test.expectedErr != nil {
+				assert.Equal(t, err, test.expectedErr)
+			}
+			assert.Equal(t, createRoleOutput, test.createIAMRoleOutput)
+		})
+	}
+}
+
+func TestCreateOrUpdateIAMPolicy(t *testing.T) {
+
+	mocks := setupDefaultMocks(t, []runtime.Object{})
+
+	mockAWSClient := mock.NewMockClient(mocks.mockCtrl)
+
+	defer mocks.mockCtrl.Finish()
+
+	awsOutputGci := &sts.GetCallerIdentityOutput{
+		Account: aws.String(""),
+	}
+
+	mockAWSClient.EXPECT().GetCallerIdentity(gomock.Any()).Return(
+		awsOutputGci,
+		nil,
+	)
+
+	mockAWSClient.EXPECT().DeletePolicy(gomock.Any()).Return(
+		&iam.DeletePolicyOutput{},
+		nil,
+	)
+
+	mockAWSClient.EXPECT().CreatePolicy(gomock.Any()).Return(
```

Comment:
Can you move this CreatePolicy EXPECT up above the DeletePolicy to better represent the order in which the calls are made?

Reply:
Please change the gomock.Any() in the DeletePolicy and CreatePolicy to better represent what we're testing - shouldn't be to big of a change as it's just passing in afr and afaa which you've defined below